enclave_build/src/pull.rs

```diff
@@ -0,0 +1,196 @@
+                })?;
+                return Ok(config_file);
+            }
+        }
+        Err(Error::DockerConfigFile(
+            "Config file not present, please set env variable \
+             DOCKER_CONFIG"
+                .to_string(),
+        ))
+    }
+}
+
+/// Returns the Docker credentials by reading from the Docker config.json file.
+///
+/// The assumed format of the file is:\
+/// {\
+///        "auths": {\
+///            "https://index.docker.io/v1/": {\
+///                    "auth": "<token_string>"\
+///            }\
+///        }\
+/// }
+pub fn parse_credentials() -> Result<RegistryAuth> {
+    let config_file = get_docker_config_file()?;
+
+    let config_json: serde_json::Value = serde_json::from_reader(&config_file)
+        .map_err(|err| Error::Credentials(format!("JSON was not well-formatted: {}", err)))?;
+
+    let auths = config_json
+        .get("auths")
+        .ok_or_else(|| Error::Credentials("Could not find auths key in config JSON".to_string()))?;
+
+    if let serde_json::Value::Object(auths) = auths {
+        for (_, registry_auths) in auths.iter() {
+            let auth = registry_auths
+                .get("auth")
+                .ok_or_else(|| {
+                    Error::Credentials("Could not find auth key in config JSON".to_string())
+                })?
+                .to_string();
+
+            let auth = auth.replace('"', "");
+            // Decode the auth token
+            let decoded = base64::decode(&auth).map_err(|err| {
+                Error::Credentials(format!("Invalid Base64 encoding for auth: {}", err))
+            })?;
+            let decoded = std::str::from_utf8(&decoded).map_err(|err| {
+                Error::Credentials(format!("Invalid utf8 encoding for auth: {}", err))
+            })?;
+
+            // Try to get the username and the password
+            if let Some(index) = decoded.rfind(':') {
+                let (username, after_user) = decoded.split_at(index);
+                let (_, password) = after_user.split_at(1);
+
+                return Ok(RegistryAuth::Basic(
+                    username.to_string(),
+                    password.to_string(),
+                ));
+            }
+        }
+    }
+
+    // If the auth token is missing, return error
+    Err(Error::Credentials("Credentials not found.".to_string()))
+}
+
+/// Determines the authentication for interacting with the remote registry.
+pub fn registry_auth() -> RegistryAuth {
+    match parse_credentials() {
+        Ok(registry_auth) => {
+            let mut _user = String::new();
+            let mut _pass = String::new();
+            println!("Credentials found.");
+            if let RegistryAuth::Basic(user, pass) = registry_auth {
+                println!("Username: {}", user);
+                println!("Password: {}", pass);
+                _user = user;
+                _pass = pass;
+            }
+            RegistryAuth::Basic(_user, _pass)
```

Comment:
the `_`s are pretty unusual. You can replace this block with:
```rust
            println!("Credentials found.");
            let (user, pass) = if let RegistryAuth::Basic(user, pass) = registry_auth {
                println!("Username: {}", user);
                println!("Password: {}", pass);
                (user, pass)
            } else {
                (String::new(), String::new())
            };
            RegistryAuth::Basic(user, pass)
```

Reply:
Also, don't use `println!` for outputing stuff. Use the `log` interface instead. And definitely do not log a `password`.

Reply:
Yes, that was a debugging leftover, sorry for that.